common/persistence/nosql/nosql_history_store_test.go

```diff
@@ -527,47 +527,630 @@ func TestDeleteHistoryBranch_unusedBranch(t *testing.T) {
+// | 	│                   │
+// | 	└ ─ ─ ─ ─ ┌ ─ ─ ─ ─ ┘
+// | 	          │
+// | 	          ┼ ─ ─ ─ ─ ── ─ ── ─ ─ ── ┐
+// | 	┌ ─ ─ ─ ─ ▼ ─ ─ ─ ─ ┐    ┌─ ─ ─ ── ┼─ ─ ─ ─ ─┐
+// | 	│                   │    │                   │
+// | 	│  <deleted>        │    │  <deleted>        │
+// | 	│                   │    │                   │
+// | 	└ ─ ─ ─ ─ ┌ ─ ─ ─ ─ ┘    └─ ─ ─ ── ┌─ ─ ─ ─ ─┘
+// | 	          │                        │
+// | 	          │                        │
+// | 	┌ ─ ─ ─ ─ ▼ ─ ─ ─ ─ ┐    ┌─ ─ ─ ── ▼─ ─ ─ ─ ─┐
+// | 	│                   │    │                   │
+// | 	│  <deleted>        │    │ <deleted>         │
+// | 	│                   │    │                   │
+// | 	└ ─ ─ ─ ─ ┌ ─ ─ ─ ─ ┘    └─ ─ ─ ── ┌─ ─ ─ ─ ─┘
+// | 	          │                        │
+// | 	          │                        │
+// | 	┌ ─ ─ ─ ─ ▼ ─ ─ ─ ─ ┐    ┌─ ─ ─ ── ▼─ ─ ─ ─ ─┐
+// | 	│                   │    │                   │
+// | 	│  <deleted>        │    │ <deleted>         │
+// | 	│                   │    │                   │
+// | 	└ ─ ─ ─ ─ ─ ─ ─ ─ ─ ┘    └─ ─ ─ ── ── ─ ─ ─ ─┘
+func TestDeleteHistoryBranch_usedBranch(t *testing.T) {
+	t.Skip("not implemented yet")
+	store, dbMock, _ := setUpMocks(t)
+
+	request := &persistence.InternalDeleteHistoryBranchRequest{
+		BranchInfo: types.HistoryBranch{
+			TreeID:   "TestTreeID",
+			BranchID: "B",
+			Ancestors: []*types.HistoryBranchRange{
+				{
+					BranchID:  "A",
+					EndNodeID: 3,
+				},
+			},
+		},
+		ShardID: testShardID,
+	}
+
+	expectedTreeFilter := &nosqlplugin.HistoryTreeFilter{
+		ShardID:  testShardID,
+		TreeID:   "TestTreeID",
+		BranchID: common.Ptr("B"),
+	}
+
+	expectedNodeFilters := []*nosqlplugin.HistoryNodeFilter{
+		{
+			ShardID:   testShardID,
+			TreeID:    "TestTreeID",
+			BranchID:  "A",
+			MinNodeID: 3,
+		},
+		{
+			ShardID:  testShardID,
+			TreeID:   "TestTreeID",
+			BranchID: "B",
+		},
+	}
+
+	dbMock.EXPECT().DeleteFromHistoryTreeAndNode(gomock.Any(), expectedTreeFilter, expectedNodeFilters).
+		Return(nil).Times(1)
+
+	historyTree := []*nosqlplugin.HistoryTreeRow{
+		{
+			ShardID:  testShardID,
+			TreeID:   "TestTreeID",
+			BranchID: "B",
+			Ancestors: []*types.HistoryBranchRange{
+				{
+					BranchID:    "A",
+					BeginNodeID: 0,
+					EndNodeID:   3,
+				},
+			},
+		},
+		// notably TestAncestorBranchID does not exist
+	}
+
+	dbMock.EXPECT().SelectFromHistoryTree(gomock.Any(), gomock.Any()).
+		Return(historyTree, nil).Times(1)
+
+	err := store.DeleteHistoryBranch(ctx.Background(), request)
+	assert.NoError(t, err)
+}
+
+// | In this scenario, there's a few branches of a normal workflow. In this example, both history_trees exist
+// | and the original branch has not yet been deleted, and (for whatever reason, the second branch is being removed
+// | before the original/parent (it's not super obvious this might happen, but it's forseeable with deletion jitter
+// | or maybe failover scenarios where this might happen).
```

Comment:
from chat: probably worth mentioning that default jitter is large so this is probably fairly common (far from a majority, but thoroughly "expected")